packages/tangle/tipmanager.go

```diff
@@ -324,9 +327,207 @@ func (t *TipManager) Tips(p payload.Payload, countParents int) (parents MessageI
+	minSupportedTimestamp := now.Add(-t.tangle.Options.TimeSinceConfirmationThreshold)
+	timestampValid = true
+
+	// skip TSC check if no message has been confirmed to allow attaching to genesis
+	if t.tangle.TimeManager.LastConfirmedMessage().MessageID == EmptyMessageID {
+		// if the genesis message is the last confirmed message, then there is no point in performing tangle walk
+		// return true so that the network can start issuing messages when the tangle starts
+		return
+	}
+
+	// if last confirmed message if older than minSupportedTimestamp, then all tips are invalid
+	if t.tangle.TimeManager.LastConfirmedMessage().Time.Before(minSupportedTimestamp) {
+		return false
+	}
+
+	if t.tangle.ConfirmationOracle.IsMessageConfirmed(messageID) {
+		// selected message is confirmed, therefore it's correct
+		return
+	}
+
+	// selected message is not confirmed and older than TSC
+	t.tangle.Storage.Message(messageID).Consume(func(message *Message) {
+		timestampValid = minSupportedTimestamp.Before(message.IssuingTime())
+	})
+	if !timestampValid {
+		// timestamp of the selected message is invalid
+		return
+	}
+
+	markerWalker := walker.New(false)
+	messageWalker := walker.New(false)
+
+	t.processMessage(messageID, messageWalker, markerWalker)
+
+	for markerWalker.HasNext() && timestampValid {
+		marker := markerWalker.Next().(markers.Marker)
+		timestampValid = t.checkMarker(&marker, messageWalker, markerWalker, minSupportedTimestamp)
+	}
+
+	for messageWalker.HasNext() && timestampValid {
+		timestampValid = t.checkMessage(messageWalker.Next().(MessageID), messageWalker, minSupportedTimestamp)
+	}
+	return timestampValid
+}
+
+func (t *TipManager) processMessage(messageID MessageID, messageWalker, markerWalker *walker.Walker) {
+	t.tangle.Storage.MessageMetadata(messageID).Consume(func(messageMetadata *MessageMetadata) {
+		if messageMetadata.StructureDetails() == nil || messageMetadata.StructureDetails().PastMarkers.Size() == 0 {
+			// need to walk messages
+			messageWalker.Push(messageID)
+			return
+		}
+		messageMetadata.StructureDetails().PastMarkers.ForEach(func(sequenceID markers.SequenceID, index markers.Index) bool {
+			if sequenceID == 0 && index == 0 {
+				// need to walk messages
+				messageWalker.Push(messageID)
+				return false
+			}
+			pastMarker := markers.NewMarker(sequenceID, index)
+			markerWalker.Push(*pastMarker)
+			return true
+		})
+	})
+}
+
+func (t *TipManager) checkMarker(marker *markers.Marker, messageWalker, markerWalker *walker.Walker, minSupportedTimestamp time.Time) (timestampValid bool) {
+	message := t.getMarkerMessage(marker)
+
+	if message.IssuingTime().Before(minSupportedTimestamp) {
+		// marker before minSupportedTimestamp
+		if !t.tangle.ConfirmationOracle.IsMarkerConfirmed(marker) {
+			// if unconfirmed, then incorrect
+			markerWalker.StopWalk()
+			return false
+		}
+
+		// if closest past marker is confirmed and before minSupportedTimestamp, then message should be ok
+		return true
```

Comment:
Why is that the case? Can you point me to a test case for this? I'm failing to see why we should stop examining the marker here. Doesn't that mean we need to walk the message past cone until we reach this marker?

Reply:
Is we process a marker that is older than TSC and is confirmed, then we can return true, as this marker is valid and it's past cone is confirmed as well. 

Actually now that I think about it, this case will never be reached as all the other conditions cover this already.
* Before we add marker to the walker, we check the oldest unconfirmed marker in the sequence, and if it's older we return immediately. 
* If tip's previous marker is older than the TSC, then we either walk messages if it's confirmed, or reject the tip if it's not confirmed.

However I'm not sure if we should remove this code as it can work as a safety measure just in case. Wdyt?

Reply:
Maybe we can add a comment about exactly this: that this "should never actually happen" and is just a safety measure.